[WIP] NUSMods Optimizer Prototyping

frizensami opened the pull request:
<!--
Thank you for contributing to NUSMods!
The template below was made to help both you and the reviewers understand
your changes. Please fill it up to the best of your ability.
(These are comments, they won't be shown in the "preview")
-->

## Context
This WIP PR accompanies the discussions on timetable optimization in #3294. 

This PR is for rapid prototyping based on the discussions in the issue + auto deployment to Vercel so that changes can be viewed easily.



## Implementation

**Note: won't try to fulfil `npm run ci` warnings/errors mostly for now, will implement functionality first**
<!-- Explain how your solution solves the problem -->
<!-- If it affects UI, an image or gif is greatly encouraged. -->

### Discussion points
Some discussion points in no particular order:
- [ ] Main entry point to view optimizer (on the main timetable? separate tab? Former seems quite natural.)
  - [ ] Should we modify the main timetable and then give users an option to undo to their initial timetable? 
- [ ] Where to display constraints? (Just below modules? Popup/Modal? Not immediately in favour of the popup/modal idea since users perhaps can't change constraints + rerun optimizer quickly) 
- [ ] How to guide users on how to use constraints? 
- [ ] When should we download the ~ 4 MB solver file? During initial NUSMods page load? Or only when the user starts to use the optimizer?
- [ ] In the future, we might have _competing_ optimization constraints, e.g., (a) maximize my free days but (b) give me lunch hours. This might require users to rank or weight their optimization preferences. 

### Implementation Plan
After further discussion I can update this PR with an implementation plan. Minimally, this needs:
- [ ] Constraints UI / UX
- [X] Integrating the Z3 worker, wasm file, and wrappers
- [x] Classes and utils to translate from an NUSMods timetable --> SMTLIB2 --> solver results --> back to a timetable
- [x] (In progress) Non-UI/UX cleanup
- [x] Properly publish the `smtlib-ext` library so we can stop the github reference
- [ ] A LOT of tests
- [x] Vendor types for Z3 and smtlib?
Of the above, the **Constraints UI** will need the most work. The latter two already have been somewhat implemented in [the standalone NUS Timetable Optimizer](https://github.com/frizensami/nusmods_smt_web/).


Translation utils:
- [ ] (In progress) NUSMods timetable --> Optimizer's GenericTimetable
- [ ] Z3 results --> list of changes to make to the existing timetable
- [ ] More to be updated

### Known bugs / limitations
- [ ] Expects only normal semesters (1, 2)
- [ ] Assumes normal weeks array (`NumericWeeks: number[]`) and not `WeekRange` (strange week ranges)
 
### Things to test
- Lesson types with spaces need to be processed before going into the solver as variable names. Solver varnames cannot have spaaces.

### Current changes 

These are just to indicate the idea, please no punterino on the UI :D


1. Implements a "Show/Hide Optimizer" button on the main timetable UI
![image](https://user-images.githubusercontent.com/5507069/123570652-43409080-d7fb-11eb-910b-5ee6305c8c0c.png)
2. Showing the optimizer (for now) displays a set of constraints below the timetable 
![image](https://user-images.githubusercontent.com/5507069/123570709-610df580-d7fb-11eb-874c-540db1c06458.png)



## Other Information
This is more or less the first time I'm using React, and I'm not much of a front-end designer (clearly, haha), so I'll try to keep up! :slightly_smiling_face: 

<!--
This section is optional, it's for stuff like:
- Any questions you may have
- Any assistance you need
- Any tasks that are incomplete
- Letting us know that you're new to this (so we know how much to help out)
- Random gifs and emojis
-->

frizensami committed:
Add optimizer button, placeholder component, state

* Optimizer button next to Exam Calendar
* Placeholder component with a hr and text
* Redux actions and reducers for maintaining optimizer state across tab changes

Redux state is necessary since we want optimizer settings to be maintained even if user navigates to check other modules

frizensami committed:
Add props to timetableoptimizer

frizensami committed:
Run format

frizensami committed:
Merge branch 'master' into optimizer

frizensami committed:
Add precompiled Z3 WASM file + Emscripten wrapper

* z3w.wasm is precompiled from https://github.com/cpitclaudel/z3.wasm/
* z3w.js is the Emscripten generated wrapper from z3w.wasm

frizensami committed:
Add Z3 WebWorker code and message passing types

frizensami committed:
Modify z3 types filename

frizensami committed:
Add callback defs to z3 types

frizensami committed:
Remove log from z3w.js

frizensami committed:
Add worker-loader package to load Z3 WebWorker

frizensami committed:
Update imports and typedefs in z3Worker

frizensami committed:
MVP for initializing Z3 with worker-loader

frizensami committed:
Fix initialization messages for z3

frizensami committed:
Add a temporary alert for z3 init

frizensami committed:
Refactor files and rename for minimal example

li-kai commented on website/package.json:
```diff
@@ -168,7 +168,8 @@
     "searchkit": "2.4.1-alpha.5",
-    "use-subscription": "1.5.1"
+    "use-subscription": "1.5.1",
+    "worker-loader": "^3.0.8"
```
This should be in devDependencies

li-kai commented on website/src/types/optimizer.ts:
```diff
@@ -0,0 +1,144 @@
+
+
+// User-selected constraints to pass to optimizer
+export interface GlobalConstraints {
```
for booleans, we try to prefix them with `isX` or `hasX` (e.g. `isWorkloadActive`)

li-kai commented on website/src/types/optimizer.ts:
```diff
@@ -0,0 +1,144 @@
+/**
+ * Defs for communicating between Optimizer <-> WebWorker <-> WASM wrapper
+ * */
+export enum Z3MessageKind {
```
Is there something stopping us from using numbers instead? Iirc, when we pass things to wasm, it is more efficient.

frizensami replied:
The Z3 message is maybe a misnomer, it's a message for the Z3 Worker. I'll rename it.

frizensami replied:
Little easier to debug with string enums, so if it's not a performance concern (very few messages are sent generally), I think the string enum might be ok.

frizensami committed:
Add sexpr-plus package for parsing Z3 output

frizensami committed:
Add smtlib-ext lib, creates SMT-LIB code for Z3

This is a library we control, provides an API for us to create SMT-LIB 2.0 code to pass to Z3.
Fork of https://github.com/stanford-oval/node-smtlib + additional APIs for our use case

frizensami committed:
Implementation of MVP optimizer utils stack

frizensami committed:
Fix incorrect callback type

frizensami committed:
Wire up button to run optimizer, fix varname bug

Optimize Timetable button runs the optimizer and prints results in console.
Realized that lesson types cannot have spaces as they are used in Z3 varnames (which cannot have space).
We convert the lesson types to their abbreviated versions to solve this issue.

frizensami committed:
Change lessons based on optimizer output

frizensami committed:
Merge branch 'master' into optimizer

frizensami committed:
Fix some eslint errors

frizensami committed:
Update worker-loader to devDependencies

frizensami committed:
Fix majority of eslint errors

frizensami committed:
Fix webworker eslint errors

frizensami committed:
Fix inline webpack loader eslint error

frizensami committed:
Correctly name constraint booleans

frizensami committed:
Z3Message -> Z3WorkerMessage for clarity

frizensami committed:
Remove unnecessary export

frizensami committed:
Comments for WeekSolver, remove StringIdGenerator

StringIdGenerator was used to create running alphabet names for variable names in Z3
This ended up being completely unnecessary, as an integer suffix for a string was sufficient.

frizensami committed:
Change slot whoId to ownerId for naming clarity

frizensami committed:
Add doc comments for timetable solver

frizensami committed:
Change SExpr to SNode in solver for accuracy

frizensami committed:
Publish smtlib-ext to remove dep on github

frizensami committed:
Add and apply vendor types for sexpr-plus

frizensami committed:
Remove all remaining eslint errors

frizensami committed:
Solve typing issues except webworker Z3

frizensami committed:
Merge branch 'master' into optimizer

frizensami committed:
Just include webworker lib

frizensami committed:
Add tests for z3weeksolver

frizensami committed:
Ignore type errors for wasm

frizensami committed:
Increase test coverage for z3 timetable solver

frizensami committed:
Add test coverage for setBooleanSelectorCosts

frizensami committed:
Autofix eslint

frizensami committed:
Full test coverage for timetable solver

frizensami committed:
Change test -> it for consistency

frizensami committed:
Add mocks and some tests for timetableOptimizer

frizensami committed:
Complete test coverage for timetableOptimizer

frizensami committed:
Fix remaining test coverage for optimizer

frizensami committed:
Ignore webworkers in codecov

frizensami committed:
Partial commit for converter

frizensami committed:
Fix errors to see converter coverage

frizensami committed:
Test coverage for converter

frizensami committed:
Merge branch 'master' into optimizer

frizensami committed:
Merge branch 'optimizer' of github.com:frizensami/nusmods into optimizer

frizensami committed:
Fix merge conflict marker
